keps/sig-network/0029-nodelocal-dns-cache.md

```diff
@@ -0,0 +1,145 @@
+
+### Non-Goals
+
+Providing a replacement for kube-dns.
+
+## Proposal
+
+A nodeLocal dns cache runs on all cluster nodes. This is managed as an add-on, runs as a Daemonset. All pods using clusterDNS will now talk to the nodeLocal cache, which will query kube-dns in case of cache misses in cluster.local domain. The node's resolv.conf will be used by this local agent for all other cache misses. 
+
+#### Daemonset, Service, Listen Interface for caching agent
+
+The caching agent daemonset runs in hostNetwork mode in kube-system namespace with a Priority Class of “system-node-critical”. It listens for dns requests on a dummy interface created on the host. A separate ip address is assigned to this dummy interface, so that requests to kube-dns or any other custom service are not incorrectly intercepted by the caching agent. This ip address is obtained by creating a nodelocaldns service, with no endpoints. Kubelet takes the service name as an argument `--localDNS=<namespace>/<svc name>`, looks up the ip address and populates pods' resolv.conf with this value instead of clusterDNS. Each cluster node will have a dummy interface with this service ip assigned to it. The Daemonset runs in privileged securityContext since it needs to create this dummy interface and add iptables rules.
+
+#### iptables NOTRACK
+
+NOTRACK rules are added for connections to and from the nodelocal dns service ip. Additional rules in FILTER table to whitelist these connections, since the INPUT and OUTPUT chains have a default DROP policy.
+
+A dnscache nanny, similar to the [dnsmasq nanny](https://github.com/kubernetes/dns/tree/master/pkg/dnsmasq) will create the dummy interface and iptables rules. The nanny gets the nodelocal dns service ip address by querying kube-dns for the service name. The nanny will also periodically ensure that the iptables rules are present. The resource usage for periodic iptables check needs to be measured. We can reduce the frequency of the checks by having the nanny query the dns cache and checking/adding rules only if the query fails.
+
+
+[Proposal presentation](https://docs.google.com/presentation/d/1c43cZqbVhGAlw3dSNQIOGuvQmDfKaA2yiAPRoYpa6iY), also shared at the sig-networking meeting on 2018-10-04
+
+Slide 5 has a diagram showing how the new dns cache fits in.
+
+#### Choice of caching agent
+
+The current plan is to run [Unbound dns server](https://www.nlnetlabs.nl/projects/unbound/about/) by default, based on these benchmark[ tests](https://github.com/kubernetes/perf-tests/tree/master/dns).
+
+Will rerun benchmark tests with coreDNS after SO_REUSEPORT support and measure QPS.
+
+Tests were run on a 1.9.7 cluster with 2 nodes, using Unbound 1.7.3 and coreDNS 1.2. Unbound QPS was 3600(caching)/3400(no caching), coreDNS 950(caching)/500(no caching)
+
+
+Based on the prototype/test results, these are the recommended defaults: 
+CPU request: 50m
```

Comment:
that's a lot more than I expected.  We'll need to see if we can rely on burstable CPU for this and once we have proper autoscaling, use it.  

@justinsb has been looking at cluster-proportional scaling again, but this would almost certainly need to be scaled based on the size of the node it lands on, which is more complicated.